Add support for executing programs

iomartin opened the pull request:
This PR adds support for executing eBPF programs on Hermes devices.

This is done either through an fsync() or by passing the RWF_SYNC flag to pwritev2().

Note that this is based on #21 and as such only the last 5 commits are actually part of this PR. I'll rebase once #21 gets in.

Also, I'm marking this as a draft PR because we don't yet have eBPF engines on the QEMU device (https://github.com/Eideticom/eid-hermes-qemu/issues/8), so I can't fully test it. But it seems to be writing/reading the right things at the right offsets. I'll change it to a regular PR once we're satisfied with https://github.com/Eideticom/eid-hermes-qemu/issues/8.

iomartin committed:
driver: execute eBPF program on an idle engine

If there are no free idle engines, we wait until one becomes available.

iomartin committed:
example: trigger program execution

We can either use fsync() or the RWF_SYNC flag on pwritev2() to start
the program.

iomartin committed:
unit_test: add tests for fsync and RWF_SYNC

Also test an invalid program

iomartin committed:
specs: describe how to execute eBPF programs

iomartin commented:
I realized I was missing a way to get the eBPF return code, so I added a couple more commits to address this. This series starts on "driver: add fsync support"

lsgunth commented on src/driver/hermes_cdev.c:
```diff
@@ -97,6 +104,80 @@ static int hermes_close(struct inode *inode, struct file *filp)
+
+	memcpy_toio(&hermes->cmds[eng].req, &cmd.req, sizeof(cmd.req));
+	iowrite8(1, &hermes->cmds_ctrl[eng]);
+
```
Certainly there must be something to wait for the program to complete here.... It won't be instant with the io write (even though it may be in qemu atm).

iomartin replied:
Good point. Do you have any suggestion on how to do this?

iomartin replied:
Probably an interrupt?

lsgunth replied:
The hardware should probably set a flag in a register when it is done, then trigger an interrupt.

The driver needs an interrupt handler and a (probably) a wait queue. When an interrupt happens it wakes up any thread waiting which would check the register for completion.

lsgunth commented on src/driver/hermes_cdev.c:
```diff
@@ -130,6 +141,12 @@ static int hermes_fsync(struct file *filp, loff_t start, loff_t end, int datasyn
 
+	eng = hermes_get_ebpf_eng(hermes);
+	if (eng == -ENOSPC)
+		return -EBUSY;
```
Probably better to design something to wait for an engine to be not busy, instead of failing in the rare case that there is no busy engine. Application writers are probably not going to get it right otherwise.

lsgunth replied:
Yeah, this isn't a great solution.

Use a wait queue. If there is no free ID,  loop forever waiting on that queue (with an interruptible wait that returns an error if it gets interrupted) until it gets an ID.

Whenever a free ID is returned to the ida, wake one sleeper.

lsgunth commented on src/driver/hermes_cdev.c:
```diff
@@ -363,6 +366,29 @@ static long hermes_download_program(struct hermes_env *env,
 }
 
+static long hermes_get_retval(struct hermes_env *env,
+		struct hermes_prog_retval_ioctl_argp *argp)
```
This isn't a good idea, it breaks support with io_uring.

We should only support return codes that are 0 or a number indicating a failure. If the failure happens we should map it into an appropriate errno and return that from fsync() or write().

iomartin committed:
driver: add fsync support

The fsync syscall can be used to trigger program execution on a Hermes
device. The program and data slots must have been previously populated.

We use a waitqueue per eBPF engine to wait for program completion. When
an engine finishes execution, the hardware must:
1. Set the EHCMDDONE register for that engine
2. Send an interrupt using the IRQ vector number for that engine.

When the interrupt is received, the wait queue is woken up and checks
the EHCMDDONE register.

Note that a bugged or malicious program that never finishes will never
cause the wake up. This can be addressed in the future by using
wake_up_interruptible_timeout, but it will need hardware support to halt
the running program (which we don't have yet).

Initially we always execute on eBPF engine 0, but this will be addressed
on another commit.

A word on the eBPF return code: we expect programs to exit with return
code 0: anything else is considered an error and the fsync returns with
-ENOEXEC. If the return code is needed, then the eBPF program should put
it somewhere in the data slot and then read() it back.

iomartin committed:
driver: support RWF_SYNC flag on pwritev2

Using this flag triggers program execution after data has been written.
It is equivalent of doing a write + fsync.

iomartin commented:
Rebased now that read_iter/write_iter are in.

I also addressed Logan's comments, except the io wait, which I'll do later and force-push again.

iomartin committed:
driver: pass hermes_pci_dev to enable_msi_msix

In order to add eBPF interrupts (in future patches), we'll need to have
access to the hermes_dev in this function.

iomartin committed:
driver: add eBPF interrupts

The eBPF engines will raise an interrupt when a program has exited.

This patch initially registers one interrupt per eBPF engine, with an
empty handler, which will be filled in future patches.

iomartin committed:
driver: move ida_wq struct

As it'll be needed in the hermes_dev struct.

iomartin committed:
driver: introduce irq_setup/irq_teardown

These are initially just wrappers around the XDMA functions, but future
patches will add the eBPF interrupt setup/teardown.

iomartin committed:
unit_test, example: use valid eBPF programs

So far the driver tests and examples were using dummy eBPF programs,
which was fine because we weren't executing them. This is no longer the
case, so we need a valid eBPF program.

So use a very simple program, that consits of just two instructions:

r0 = 0 # opcode 0xb7
ret    # opcode 0x95

iomartin commented:
A lot has changed since the last review and a new review would be appreciated. The eBPF unit tests are still a bit of work in progress but I'd appreciate some early feedback.

lsgunth commented on src/driver/hermes_mod.c:
```diff
@@ -55,14 +55,74 @@ static const struct pci_device_id pci_ids[] = {
+struct ebpf_irq_arg {
+	struct hermes_dev *hdev;
+	int eng;
+} ebpf_irq_args[EBPF_ENG_NUM_MAX];
```
This global variable seems highly questionable... What happens if there are multiple devices in the system?

Should really have a per-engine structure that's allocated at init for this kind of thing. Then there's no need for the silly MAX define.

iomartin replied:
Yes... that makes total sense...

lsgunth commented on src/driver/hermes_cdev.c:
```diff
@@ -97,6 +104,105 @@ static int hermes_close(struct inode *inode, struct file *filp)
+static int __cmd_comp(struct hermes_dev *hermes, int eng)
+{
+	union hermes_cmd_ctrl ctrl;
+	ctrl.ehcmdctrl = ioread8(&hermes->cmds_ctrl[eng].ehcmdctrl);
```
Why is this a crazy union?

Can't it just be `return ioread8()`?

iomartin replied:
The spec currently assigns each bit of the EHCMDCTRL register a different meaning https://github.com/Eideticom/eid-hermes/blob/master/specs/eid-hermes-interface.md#ehcmdctrl

Hence the bit field in the other comment... But I guess we have plenty of space in the BAR so that both EHCMDEXEC and EHCMDDONE can take a byte each. I'll submit a separate PR for that

lsgunth replied:
It can be a bit field just fine.... But then you should really use a & MASK or some such. I still don't understand why it's a union.

lsgunth commented on src/driver/hermes_mod.h:
```diff
@@ -79,6 +89,47 @@ struct __attribute__((__packed__)) hermes_cfg {
+	union {
+		struct __attribute__((__packed__)) {
+			uint64_t ebpf_ret;
+		} run_prog;
```
Do we really need this abstraction at this point? Do we ever have plans to create non-ebpf program runners? Are you sure such things will not require the same arguments? I wouldn't predict the future and skip the union. Adding some reserved space is a good idea, but no need to do much else.

iomartin replied:
Fair point. I'll remove it from the next iteration.

lsgunth commented on src/driver/hermes_mod.h:
```diff
@@ -79,6 +89,47 @@ struct __attribute__((__packed__)) hermes_cfg {
+union __attribute__((__packed__)) hermes_cmd_ctrl {
+	struct {
+		uint8_t ehcmdexec:1;
+		uint8_t ehcmddone:1;
```
I would avoid bit fields like the plague especially for hardware interfaces. Nothing guarantees the bit order.

lsgunth commented on src/driver/hermes_cdev.c:
```diff
@@ -247,14 +246,20 @@ static ssize_t hermes_read_write_iter(struct kiocb *iocb, struct iov_iter *to)
 		xdma_release_c2h(chnl);
 
-	return res;
+	if (write && iocb->ki_flags & IOCB_SYNC) {
```
If xdma_channel_read_write() fails, we probably should not do the fsync()....

iomartin replied:
Ops, nice catch

lsgunth reviewed (commented):
There's a lot to review here.... I glossed over the last few patches.

iomartin committed:
driver: keep track of the program length

The program length is specified when we do an ioctl to DMA to a program
slot. This will be necessary for when we create a command request.

sbates130272 reviewed (approved):
@iomartin this looks very nice. I did not walk through this line by line but at a high level you have my approval.

iomartin commented:
Now that https://github.com/Eideticom/eid-hermes-qemu/pull/14 is getting closed to being merged, it'd be great if we could get this one in too, as one depends on the other.

niclashedam commented:
PR #43 and PR #44 should be merged before we merge this.

I have deployed the fsync code to a physical device and tested the implementation there. Everything works beside #43 and #44, so I believe it is a good time to merge this into master. Can you allocate some time for merging this @sbates130272 @iomartin ?

sbates130272 commented:
@iomartin Is this ready to be merged?

iomartin commented:
@sbates130272 I think we're close, I just want to clear the two PRs that @niclashedam opened as they're potential bugfixes that should be addressed with this PR
